### DISCUSSION TITLE
Using Sys.Argv's With Pytest

### DISCUSSION THREAD

**Discussion by DevishaPadmaperuma** [pytest-dev/pytest#8783] [Q&A] [answered]
Hi Everyone,

Im a beginner at python, I wanted to make a program that'll take sys_args and generate a password.
I wanted to use pytest to debug it, through github's workflow. (this is one file, the other's for flask). And i tried the following
`pytest salt.py make 1` but it says;

`================================================== no tests ran in 0.02s ===================================================
ERROR: file or directory not found: make`

Here's my code:

```python
def main():
    proceed = sys.argv[1]
    if proceed == 'make':
        count = int(sys.argv[2])
        make_name(count)
    else:
        ti= time.time()
        name = sys.argv[2]
        count = sys.argv[3]
        if count == 'all':
            count = 'all'
            make_salt(count, name)
        else:
            make_salt(count, name)
        print(f"Time Taken: {round(time.time() - ti, 2)}s")
```

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Pytest is a own cli program with own cli handling

If you want to test a cli you have to do what everyone else is doing and enable your main function to take an alternative to sys.argv

I can elaborate once I get back to the computer later if necessary

**Reply by DevishaPadmaperuma** [reply]
Hi @RonnyPfannschmidt ,
Thanks for the reply. I'm not entirely sure on how to do that, so could you please elaborate whenever you're free.
Thanks!

**Reply by RonnyPfannschmidt** [reply]
a common pattern is
```python
def main(args=None):
    if args is None:
        args = sys.argv[1:]
    ....
```

then other code can just call 
```
main(["make","all"])
```

**Reply by DevishaPadmaperuma** [reply]
It worked, thanks alot!

**Reply by hughesadam87** [reply]
I wrote a small blog post on this topic that provides a miniimal working example for the above and how to test

https://hughesadam87.medium.com/dead-simple-pytest-and-argparse-d1dbb6affbc3
